homeassistant/components/wled/coordinator.py

```diff
@@ -33,9 +39,53 @@ def update_listeners(self) -> None:
         for update_callback in self._listeners:
             update_callback()
 
+    async def _use_websocket(self) -> None:
+        """Use WebSocket for updates, instead of polling."""
+        try:
+            await self.wled.connect()
+        except WLEDError as err:
+            self.logger.info(err)
+            return
+
+        async def close_websocket(_) -> None:
+            """Close WebSocket connection."""
+            await self.wled.disconnect()
+
+        # Clean disconnect WebSocket on Home Assistant shutdown
+        self.unsub = self.hass.bus.async_listen_once(
+            EVENT_HOMEASSISTANT_STOP, close_websocket
+        )
+
+        async def listen() -> None:
+            """Listen for state changes via WebSocket."""
+            try:
+                await self.wled.listen(callback=self.async_set_updated_data)
+            except WLEDConnectionClosed as err:
+                self.last_update_success = False
+                self.logger.info(err)
+            except WLEDError as err:
+                self.last_update_success = False
+                self.update_listeners()
+                self.logger.error(err)
+
+            # Ensure we are disconnected
+            await self.wled.disconnect()
+            if self.unsub:
+                self.unsub()
+                self.unsub = None
+
+        # Start listening
+        asyncio.create_task(listen())
+
     async def _async_update_data(self) -> WLEDDevice:
         """Fetch data from WLED."""
         try:
-            return await self.wled.update(full_update=not self.last_update_success)
+            device = await self.wled.update(full_update=not self.last_update_success)
         except WLEDError as error:
             raise UpdateFailed(f"Invalid response from API: {error}") from error
+
+        # If the device supports a WebSocket, try activating it.
+        if device.info.websocket is not None and not self.wled.connected:
```

Comment:
Could this be a race condition when a user calls `homeassistant.update_entitiy` and it hits this method 3 times, and then creates 3 tasks to start the WS? 

If you make the `use_websocket` method a callback, and then check `self.unsub`, you won't have this problem.

Reply:
Done 👍